src/clang_parser.cpp

```diff
@@ -584,46 +619,50 @@ bool ClangParser::parse(ast::Program *program, BPFtrace &bpftrace, std::vector<s
     args.push_back("-D__CLANG_WORKAROUNDS_H");
   }
 
-  auto incomplete_types = get_incomplete_types(input, input_files, args);
-  bpftrace.btf_set_.insert(incomplete_types.cbegin(), incomplete_types.cend());
+  // The generated BTF header is initially empty
+  std::string btf_cdef;
+  input_files.emplace_back(CXUnsavedFile{
+      .Filename = "/bpftrace/include/__btf_generated_header.h",
+      .Contents = btf_cdef.c_str(),
+      .Length = btf_cdef.size(),
+  });
 
   ClangParserHandler handler;
-  CXErrorCode error;
-
-  if (process_btf)
+  bool check_additional_types = true;
+  while (check_additional_types)
   {
-    auto btf_and_input = "#include <__btf_generated_header.h>\n" + input;
-    auto btf_and_input_files = getTranslationUnitFiles(
-        CXUnsavedFile{ .Filename = "definitions.h",
-                       .Contents = btf_and_input.c_str(),
-                       .Length = btf_and_input.size() });
-
-    auto btf_cdef = bpftrace.btf_.c_def(bpftrace.btf_set_);
-    btf_and_input_files.emplace_back(CXUnsavedFile{
+    auto incomplete_types = get_incomplete_types(
+        input, input_files, args, bpftrace.btf_set_);
+    unsigned types_cnt = bpftrace.btf_set_.size();
+    bpftrace.btf_set_.insert(incomplete_types.cbegin(),
+                             incomplete_types.cend());
+
+    if (process_btf)
+    {
+      // Update generated header with current BTF types
+      btf_cdef = bpftrace.btf_.c_def(bpftrace.btf_set_);
+      input_files.back() = CXUnsavedFile{
         .Filename = "/bpftrace/include/__btf_generated_header.h",
         .Contents = btf_cdef.c_str(),
         .Length = btf_cdef.size(),
-    });
-
-    error = handler.parse_translation_unit(
-        "definitions.h",
-        args.data(),
-        args.size(),
-        btf_and_input_files.data(),
-        btf_and_input_files.size(),
-        CXTranslationUnit_DetailedPreprocessingRecord);
-  }
-  else
-  {
-    error = handler.parse_translation_unit(
-        "definitions.h",
-        args.data(),
-        args.size(),
-        input_files.data(),
-        input_files.size(),
-        CXTranslationUnit_DetailedPreprocessingRecord);
+      };
+      // If additional BTF types were found, we need to repeat the process since
+      // that might have introduced some new unresolved typedefs.
+      check_additional_types = types_cnt != bpftrace.btf_set_.size();
```

Comment:
Not really, we need to check if a **new** incomplete type was found (one that wasn't in `bpftrace.btf_set_` already), otherwise the loop could not terminate (see the comment above for example).